cpp/src/parquet/arrow/arrow_reader_writer_test.cc

```diff
@@ -3719,6 +3719,77 @@ TEST(TestArrowReaderAdHoc, WriteBatchedNestedNullableStringColumn) {
   ::arrow::AssertTablesEqual(*expected, *actual, /*same_chunk_layout=*/false);
 }
 
+TEST(TestArrowReaderAdHoc, RepeatReadNullableStructColumnFile) {
+  // ARROW-14047
+  auto pool = default_memory_pool();
+
+  // Special parquet file; see attachment in Jira
+  auto file_path = test::get_data_file("writeReadRowGroup.parquet");
+  PARQUET_ASSIGN_OR_THROW(auto infile, ::arrow::io::ReadableFile::Open(file_path, pool));
+
+  std::unique_ptr<parquet::arrow::FileReader> arrow_reader;
+  ASSERT_OK(OpenFile(infile, pool, &arrow_reader));
+
+  std::shared_ptr<::arrow::Table> first_read;
+  ASSERT_OK(arrow_reader->ReadTable(&first_read));
+
+  ASSERT_OK(first_read->column(1)->chunk(0)->ValidateFull()); // So it's in theory valid
+
+  std::shared_ptr<::arrow::Table> second_read;
+  ASSERT_OK(arrow_reader->ReadTable(&second_read));
+  ASSERT_OK(arrow_reader->ReadTable(&second_read)); // (Actually third read)
+
+  // 'first_read->column(1)->chunk(0)->ValidateFull()' failed with Invalid: List child array invalid: Invalid: Struct child array #0 invalid: Invalid: null_count value (854) doesn't match actual number of nulls in array (861)
+  // ASSERT_OK(first_read->column(1)->chunk(0)->ValidateFull()); 
+
+  ::arrow::AssertTablesEqual(*first_read.get(), *second_read.get());
+
+  // To eliminate possibility of parquet writer oddities, rewrite table to new parquet:
+  using ::arrow::io::BufferOutputStream;
+  ASSERT_OK_AND_ASSIGN(auto outs, BufferOutputStream::Create(1 << 10, pool));
+  auto props = default_writer_properties();
+  std::unique_ptr<arrow::FileWriter> writer;
+  ASSERT_OK(
+      arrow::FileWriter::Open(*first_read->schema().get(), pool, outs, props, &writer));
+  ASSERT_OK(writer->WriteTable(*first_read.get(), std::numeric_limits<int64_t>::max()));
+  ASSERT_OK(writer->Close());
+  ASSERT_OK_AND_ASSIGN(auto buffer, outs->Finish());
+
+  // Read from table
+  std::unique_ptr<parquet::arrow::FileReader> arrow_reader2;
+  ASSERT_OK(parquet::arrow::OpenFile(std::make_shared<BufferReader>(buffer), pool, &arrow_reader2));
+
+  ASSERT_OK(arrow_reader2->ReadTable(&first_read));
+  ASSERT_OK(first_read->column(1)->chunk(0)->ValidateFull()); // So it's in theory valid
+
+  ASSERT_OK(arrow_reader2->ReadTable(&second_read));
+  // 'first_read->column(1)->chunk(0)->ValidateFull()' failed with Invalid: List child array invalid: 
+  // Invalid: Struct child array #0 invalid: Invalid: null_count value (854) doesn't match actual number 
+  // of nulls in array (861)
+  // ASSERT_OK(second_read->column(1)->chunk(0)->ValidateFull()); 
+
+  ::arrow::AssertTablesEqual(*first_read.get(), *second_read.get());
```

Comment:
So this line fails?

Reply:
Yes, if you comment out line 3739. As is, it will error on line 3763 (ValidateFull). (Sorry that's a little confusing)

Output from error on 3771:

```
56: /Users/willjones/Documents/arrows/arrow/cpp/src/arrow/testing/gtest_util.cc:205: Failure
56: Failed
56: # chunk 0
56: 
56: @@ -0, +0 @@
56: -[{}, {}, {}, {}, {}, {}, {}]
56: -[{}, {}, {x: 13}, {}, {x: 13.2}, {}, {x: 13.4}, {}]
56: +[{x: 12.7}, {x: 12.8}, {}, {x: 13}, {x: 13.1}, {x: 13.2}, {x: 13.3}]
56: +[{x: 12.8}, {}, {x: 13}, {}, {x: 13.2}, {}, {x: 13.4}, {}]
```

Reply:
So if I read this correctly there is an list<struct<x: double>> arrow schema we are trying to read back?

Reply:
Yes.

And I just found the resulting array is invalid in every other read starting with the third. That is the following reads (1-indexed) will be invalid: 3, 5, 7, 9, 11, 13, 15, 17, 19, 21, 23, 25, 27, 29,

Reply:
So reads 0,1,2 and then and all even reads are valid? do they all agree with each other?

Reply:
That's what I'm seeing so far. And I believe the original test showed the valid ones do, but I will do another check.

Reply:
Yes, all evens are equal.

Reply:
Wow, that is strange.  Also, just to check they also equal after round-tripping or is there data loss?  Most of the state should be constructed by scratch on read, so I can't think of anything off the top of my head that would cause this.

Reply:
Thanks for looking at it, Micah. I'll try ASAN and UBSAN, as well as see if I can repro with fuzz testing.